### ISSUE TITLE
Problema de renderização/duplicação do Portugol Studio (Soluções e Discussão)

### ISSUE THREAD

**Issue by AdsonEsteves** [UNIVALI-LITE/Portugol-Studio#421]
# Fio de resolução do problema de renderização/duplicação do Portugol Studio

## O Problema

Existe um problema que acontece em alguns computadores onde o Portugol Studio (como também algumas outras aplicações JAVA) começa a se duplicar pela tela do usuário, sem nenhum padrão muito específico, sendo muito difícil de replicar, pois aparentemente são algumas especificações da versão do sistema, placa de vídeo e ainda problemas do próprio JAVA.

Exemplo de imagem com problema ocorrendo:
![image](https://user-images.githubusercontent.com/21344756/30877154-3ed548b6-a2cf-11e7-96e7-68817831edaa.png)

Como já foi visto e discutido nas issues #80 #231 #371 #417 , além de alguns fios no StackOverFlow como  [este](https://stackoverflow.com/questions/34316000/swing-ui-multiplying-panel-ghosting) e [este outro](https://stackoverflow.com/questions/22737535/swing-rendering-appears-broken-in-jdk-1-8-correct-in-jdk-1-7), este é um problema que ocorre em computadores com sistemas em versões específicas e/ou placas de video com versões específicas.

### 1. Atualizar

A maioria das vezes o problema foi resolvido apenas atualizando o sistema ou o driver da placa de video.
Por isso se há alguma atualização do seu Sistema Operacional pendente ou uma atualização da sua placa de video tente atualiza-los antes de mais nada.

### 2. Variável de Ambiente

Existe a possibilidade de ele poder ser possivelmente resolvido desativando o Direct3D que o JAVA usa. Aqui pretendo apresentar o método de desativá-lo.

1. Entre nas variáveis de ambiente do seu sistema
  - abra o executar (![](https://i.stack.imgur.com/T0oPO.png)+R)
  - escreva ``control sysdm.cpl,,3`` e dê ok
  - na janela aberta clique em Variáveis de ambiente

2. Coloque as variáveis `sun.java2d.d3d` e `J2D_D3D`(se não tiver crie novas) e coloque ambas como `false`. 
- Deve ficar como nas imagens abaixo:
![image](https://user-images.githubusercontent.com/8836540/32119271-745a8c18-bb33-11e7-95f6-dca7c719ea91.png)
![image](https://user-images.githubusercontent.com/8836540/32119522-4cf1ddd8-bb34-11e7-82d8-d927b94ea1ab.png)

- Deve terminar assim:
![image](https://user-images.githubusercontent.com/8836540/32119656-b222a124-bb34-11e7-98a3-38711de46c1f.png)
![image](https://user-images.githubusercontent.com/8836540/32119670-bc6de4cc-bb34-11e7-8d2f-2e5b2e6a461d.png)

## Por fim

Se nenhuma das soluções funcionou, apresente aqui as especificações do seu sistema para futura referência, se possível neste formato
```
Windows 10 Education
Versão: 1703
Build: 15063.608
Processador: AMD FX-8300
8Gb Ram
Vídeo: GTX 1050 Ti 4gb
Resolução da tela: 2560x1080
```

E caso tenha conseguido resolver de alguma outra maneira, nos mostre aqui, para que outras pessoas também consigam.

(ﾉ◕ヮ◕)ﾉ*:･ﾟ✧ Obrigado ✧ﾟ･: *ヽ(◕ヮ◕ヽ)

**Comment by danielthejoker** [comment]
Resolveu o problema pra mim, valeu!

**Comment by MackelDuarte** [comment]
Resolvel aqui muito obg,

**Comment by noschang** [comment]
@AdsonEsteves, surgiu uma nova pista sobre esse bug. O prof. André Raabe recebeu o seguinte email:

> o erro estava no fato de que eu por possuir uma placa de video da MSI gtx1080ti uso o Afterburner junto com o RIVA como programas de monitoração dessa gpu, e esses programas bugam o Portugol Studio. Fechei os programas e deu certo, parou de bugar. Agora já não abrem mais janelas sobrepostas uma as outras sem permissão ou requisição. O problema foi resolvido.

Esses dois programas que ele falou, o RivaTuner e o MSI Afterburner, são softwares que interceptam as chamadas de desenho da GPU pra exibir a contagem de FPS na tela. Devem estar interferindo.

Consegues fazer alguns testes instalando esses softwares e vendo como o PS se comporta?

**Comment by AdsonEsteves** [comment]
@noschang 

É uma possibilidade. Mas esses programas não funcionariam apenas nas placas MSI?

**Comment by noschang** [comment]
@AdsonEsteves não. Eles funcionam em qualquer placa compatível com OpenGL e DirectX, ou seja, praticamente todas

**Comment by BL4CKH47** [comment]
Manos, o exemplo do amigo funcionou pra mim. Meu MSI afterburner estava configurado pra iniciar junto com o windows... No painel de controle msm eu finalizei ele e o Portugol funcionou corretamente.Acredito que o @noschang esteja certo. Não sei. De qualquer forma não funcionou no pc da minha esposa que de diferença pro meu setup é so a placa de vídeo. 
Ela está usando o navegador e ta dando certo.

**Comment by kelwingarcia** [comment]
Use a opção '2. Variável de Ambiente'
e ate agora parece estar funcionando.
Windows 10 Pro
versão 1909
Processador Intel i5-8400
RAM 32,0GB
Placa de Vídeo Geforce RTX 2060

**Comment by noschang** [comment]
@AdsonEsteves estava pensando sobre esse problema aqui. Não acha que seria interessante criar uma solução dentro do próprio PS para essa issue em específico?

Por exemplo, em algum lugar da interface colocaríamos um botão ou link escrito "A tela do Portugol Studio está duplicando? Clique aqui para ver possíveis soluções".

Ao clicar nesse botão, abriria uma tela sugerindo a primeira resolução possível e duas opções para o usuário selecionar: 
1) Esta solução resolveu meu problema
2) Esta solução não resolveu meu problema

Se o usuário selecionar a primeira opção, só fechamos a janela
Se o usuário selecionar a segunda opção, a tela é atualizada e exibe a segunda solução possível, e assim por diante até esgotar as soluções possíveis

**Comment by AdsonEsteves** [comment]
@noschang 
Pode ser. Como não tem como identificar esse erro pela programação, talvez então colocar no "Reportar um bug", já que é por onde se acessa o git geralmente (acho).
Só teria que pensar em uma maneira que não ficasse estranho, fora de contexto/padrão.

**Comment by noschang** [comment]
> @noschang
> Pode ser. Como não tem como identificar esse erro pela programação, talvez então colocar no "Reportar um bug", já que é por onde se acessa o git geralmente (acho).
> Só teria que pensar em uma maneira que não ficasse estranho, fora de contexto/padrão.

@AdsonEsteves acho a ideia ótima. Poderia acrescentar um botão na tela de reporte de erro, mais ou menos como na figura abaixo. Ao clicar no botão o PS abriria esse passo a passo que eu sugeri

![Captura de tela de 2020-11-08 00-34-04](https://user-images.githubusercontent.com/709553/98456315-ff5d2600-215a-11eb-9c96-49b8d36db264.png)

**Comment by MarcuusVRCD** [comment]
A Opção 2 funcionou comigo. Era o MSI Afterburner. Obrigado pela ajuda.

**Comment by pedro-hms** [comment]
No meu caso eu só precisei fechar o MSI Afterburner e essa falha deixou de existir. 

Windows 11 PRO - Build 22000.120

APU AMD A10-9600P + AMD RADEON R7 M340
RAM 12GB

**Comment by thigothug** [comment]
Quanto eu abro o Portugol Studio aqui, a tela fica piscando, só quando eu to no programa em execução, fora dele fica normal, uso  msi afterburn, mas li comentarios acima e fechei, reinicie o programa e mesmo assim a tela fica piscando quando eu com ele em execução, alguma ajuda referente a isso? 

Win 10 Pro
Versão 21H1
Ryzen 7 3700x
16GB 3000mhz
Gtx 1660 Super

**Comment by AdsonEsteves** [comment]
@thigothug 
Primeira vez que ouvimos falar desse problema. Particularmente não sei o que pode ser e como não temos como verificar no momento, vai ser demorada uma resposta a isso.
O que sugiro é verificar se não existe nenhum outro tipo de programa rodando que possa afetar os gráficos do sistema.

**Comment by raphabaroli** [comment]
Fiquei desesperado kkkkkk No meio da aula e eu com o portugol bugado. No fim era o Rivaturner ativado bugando o portugol... Oshe

**Comment by EdgarBrabo** [comment]
resolvi fechando o afterburner

**Comment by rafaolivn** [comment]
Win 11 home
AMD Ryzen 5 5600H with Radeon Graphics            3.30 GHz
RTX 3050

a terceira opção deu certo para mim, tive que reiniciar para funcionar no portugol.

**Comment by Celito12** [comment]
Boa tarde, aqui p mim resolveu apenas fechando o MSI Afterburner.
